default/python/universe_generation/empires.py

```diff
@@ -59,68 +58,212 @@ def get_starting_species_pool():
-    while attempts > 0:
-        attempts -= 1
-        # shuffle our pool of systems so each try the candidates are tried in different order
-        # (otherwise each try would yield the same result, which would make trying several times kind of pointless...)
-        random.shuffle(systems_pool)
-        # try to find systems that meet our condition until we either have the requested number
-        # or we have tried all systems in our pool
+    # store the result of all attempts in a list, so we can pick the attempt that yielded the largest set of systems
+    accepted_list = []
+    while attempts:
         accepted = []
-        for candidate in systems_pool:
-            # check if our candidate is at least min_jumps away from all other systems we already found
-            if (all(fo.jump_distance(candidate, system) >= min_jumps for system in accepted) and
-                    ((additional_reqs is None) or additional_reqs(candidate))):
-                # if yes, add the candidate to the list of accepted systems
-                accepted.append(candidate)
-                # if we have the requested number of systems, we can stop and return the systems we found
-                if len(accepted) >= num_systems:
-                    return accepted
-    # all tries failed, return an empty list to indicate failure
+        # get candidates that are at least min_jumps away from the systems we already accepted
+        for candidate in systems_min_jumps_away_from(systems_pool, min_jumps, accepted):
+            accepted.append(candidate)  # add the candidate to our list of accepted systems
+            # if the number of home systems has been specified and we met that number, we got what we need
+            # we can stop and return the result of the current attempt immediately
+            if (num_systems > 0) and (len(accepted) >= num_systems):
+                return accepted
+        # add the result of the current attempt to our list and continue with next attempt
+        accepted_list.append(accepted)
+        attempts -= 1
+    # return the result of the attempt that yielded the most systems
+    return max(accepted_list, key=len)
+
+
+def find_home_systems(num_systems, complete_pool, preferred_pool, min_jumps):
+    """
+    Tries to find a specified number of home systems which are as far apart from each other as possible.
+    Starts with the specified minimum jump distance and reduces that limit until enough systems can be found.
+    For each minimum jump distance two attempts are made: the first attempt uses a pool of preferred systems,
+    if that fails, a second attempt is made with a complete pool that should contain more systems (and therefore
+    has a greater chance of success).
+    """
+    # try to find home systems, decrease the min jumps until enough systems can be found, or the min jump distance
+    # gets reduced to 0 (meaning we don't have enough systems to choose from at all)
+    while min_jumps > 0:
```

Comment:
> You can save one line by replacing while to for min_jumps in reversed(range(min_jumps):

Yeah, I know, but I deliberately choose the `while` loop here. IMO that's slightly more readable code in this case, and anyway, using a `for` loop with a range may save one line of code, but that doesn't mean the resulting code is simpler (as the `while` loop can do without creating the extra sequence). So my personal preference in such cases is using `while` loops.

> To remove code duplication you can iterate over list

Good idea, rewrote the function accordingly.